stdlib/Statistics/test/runtests.jl

```diff
@@ -474,8 +474,16 @@ end
     @test quantile([0,1],1e-18) == 1e-18
     @test quantile([1, 2, 3, 4],[]) == []
     @test quantile([1, 2, 3, 4], (0.5,)) == (2.5,)
-    @test quantile([4, 9, 1, 5, 7, 8, 2, 3, 5, 17, 11], (0.1, 0.2, 0.4, 0.9)) == (2.0, 3.0, 5.0, 11.0)
+    @test quantile([4, 9, 1, 5, 7, 8, 2, 3, 5, 17, 11],
+                   (0.1, 0.2, 0.4, 0.9)) == (2.0, 3.0, 5.0, 11.0)
+    @test quantile(Union{Int, Missing}[4, 9, 1, 5, 7, 8, 2, 3, 5, 17, 11],
+                   [0.1, 0.2, 0.4, 0.9]) == [2.0, 3.0, 5.0, 11.0]
+    @test quantile(Any[4, 9, 1, 5, 7, 8, 2, 3, 5, 17, 11],
+                   [0.1, 0.2, 0.4, 0.9]) == [2.0, 3.0, 5.0, 11.0]
     @test quantile([1, 2, 3, 4], ()) == ()
+    @test isempty(quantile([1, 2, 3, 4], Float64[]))
+    @test quantile([1, 2, 3, 4], Float64[]) isa Vector{Float64}
+    @test quantile([1, 2, 3, 4], []) isa Vector{Any}
```

Comment:
Could be worth testing with `p::Vector{Any}` when it is non-empty, since we allow for it.

Reply:
Good point. I have added two more tests. What is relevant here is that using `map` we have some type instability (this problem was present with tuple version already) - I expose this instability in tests.

Here is the problem (I show it on tuple which works the same before and after this PR):
```
julia> quantile([4, 9, 1, 5, 7, 8, 2, 3, 5, 17, 11],
                          (0.1, 0.2, 0.4, 0.9))
(2.0, 3.0, 5.0, 11.0)

julia> quantile(Any[4, 9, 1, 5, 7, 8, 2, 3, 5, 17, 11],
                          (0.1, 0.2, 0.4, 0.9))
(2, 3, 5, 11)

julia> quantile([4, 9, 1, 5, 7, 8, 2, 3, 5, 17, 11],
                          (0.1, 0.2, 0.4, 0.99))
(2.0, 3.0, 5.0, 16.400000000000002)

julia> quantile(Any[4, 9, 1, 5, 7, 8, 2, 3, 5, 17, 11],
                          (0.1, 0.2, 0.4, 0.99))
(2, 3, 5, 16.400000000000002)
```

Reply:
Ah, that's annoying. Maybe change:
```julia
T = promote_type(eltype(v), typeof(v[1]*h))
```
to
```julia
T = promote_type(typeof(v[1]), typeof(v[1]*h))
```
Or maybe we can just do `T = typeof(v[1]*h)`? What's the point of using promotion?

Reply:
Both proposals are not ideal as `v[1]` may have a type unrelated to the type we actually need, e.g. for `Vector{Real}` with heterogeneous inputs. I have proposed something else that I think better captures what we want - it always combines the types of elements required and the type of quantile required.

Note that in corner cases of `0` or `1` as quantile we still get an integer:
```
julia> quantile([1,2,3], 1)
3

julia> quantile([1,2,3], 0)
1
```
but I think it is OK to leave it as is (we could enforce these cases to produce a float result, but I am not sure if we should, as in general we do not guarantee that `quantile` returns an `AbstractFloat`).

Reply:
Makes sense. Just to be sure: when you say we get an integer for `0` and `1`, that's not the case with `0.0` and `1.0`, right?

Reply:
Right. Simply only if you pass an integer as a quantile (which is possible if the quantile is equal to `0` or `1`). The key line is `f0 = (lv - 1)*p` with evaluates to an integer if `p` is an integer, and this propagates (as we only have additions and multiplications later).